### PR TITLE
IDE & Docker Integration

### LINKED ISSUE #4515: IDE and docker integration
#### Expected Behavior

- Add support to debug and run QCAlgorithms, developing locally in an IDE (VS Code, pycharm, visual studio, etc) and executing in Leans docker container. Add required documentation, scripts and configuration files for each IDE
#### Actual Behavior

- Currently, we have `run_docker.sh/bat`, which don't allow providing an external algorithm to run `.py/.dll` (nor debugging)
#### Potential Solution

Implement 1 IDE at the time, first running a custom algorithm, second debugging.
#### Reproducing the Problem

N/A
#### System Information

N/A
#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by C-SELLERS**
#### Description

This PR addresses issue #4515 

- Addition of using external algorithms inside of our docker container (Python or C#)
Done via mounting python algorithm folder as well as C# dll to docker.

- Integration of using the docker image with VS Code, Visual Studio, & PyCharm for debugging LEAN in the container

- New Python debugging configurations using PTVSD & Pycharm's PyDevD, implemented both in DebuggerHelper, works with local and remote instances of Lean.

- C# Debugging with mono implemented, works remotely and locally.

#### Related Issue

Closes #4515 

#### Motivation and Context

The usage of the container was limited by algorithms already in the container image. Plus no means of debugging existed for the process inside of the container.

#### Requires Documentation Change

Documentation has been added to the .vscode directory and .idea directory as readme files

We may want to update our debugging guide [here](https://www.quantconnect.com/tutorials/open-source/debugging-python). It actually already has a way to use PTVSD but it was setup to be used in their algorithms and now it is built into Lean and works with both VS Code and Visual Studio.

We should also add the pycharm setup which is covered in the .idea readme.

#### How Has This Been Tested?

Tests:
- Creating new algorithms and using the scripts to attach them to the docker image. Both the .bat and .sh file are successful in mounting and running Python and C# algorithms.
- Using the new VS Code tasks and launch configurations such as _launch_, _build_, and _debug in container_ to ensure it is possible to develop in VS code and debug in the container. 
- Tested mono attach in VS and VS Code for C# debugging
- Tested Pycharm debugging config for python remote and local debugging using Pycharms pydevd.
- Tested using the PTVSD configuration for lean, and attaching to it remotely and locally on port 5678 

All of this was tested in both windows and linux environment.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [X] I have added tests to cover my changes. 
- [X] All new and existing tests passed.
- [X] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by C-SELLERS** [comment]
Added debugging configuration for Python files in the container, also works for debugging python locally.

Requires PTVSD python module, has been added to Dockerfile as well.

**Comment by C-SELLERS** [comment]
Squashed the commits down

File: .vscode/tasks.json [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            "label": "run-docker",
+            "type": "shell",
+            "isBackground": true,
+            "windows": {
+                "command": "${workspaceFolder}/run_docker.bat"
+            },
+            "linux": {
+                "command": "${workspaceFolder}/run_docker.sh"
+            },
```
Comment (danoe) [inline .vscode/tasks.json:48]: could you add osx
```
            "osx": {
                "command": "${workspaceFolder}/run_docker.sh"
            },
´´´

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            "name": "Attach to Python (Container)",
+            "type": "python",
+            "request": "attach",
+            "port": 5678,
+            "pathMappings":[{ 
+                "localRoot": "${workspaceFolder}", 
+                "remoteRoot": "/Lean/"
+            }]
+        }
```
Comment (danoe) [inline .vscode/launch.json:49]: missing ,

**Comment by C-SELLERS** [comment]
Mono C# debugging is no longer working on my local machine. After a lot of troubleshooting I believe this has to do with my build configuration but I have not been able to pinpoint exactly what it is.

As a result I have build some of the mono tasks from the VS Code configuration since I cannot test them fully.

File: .dockerignore [hunk 3]
```diff
@@ -1,6 +1,4 @@
...
-.git/*
-.github/*
-.vs/*
-.nuget/*
-Tests/*
\ No newline at end of file
+.git/
+.github/
+.vs/
+.nuget/*
```
Comment (Martin-Molinero) [inline .dockerignore:4]: Wonder if removing `packages` and `Tests` from being ignored is a required change?

File: .idea/readme.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+
+
+
+    *   Windows
+        1. [https://docs.docker.com/docker-for-windows/#file-sharing](https://docs.docker.com/docker-for-windows/#file-sharing)
+        2. Share the LEAN root directory with docker
+    *   Mac
+        3. [https://docs.docker.com/docker-for-mac/#file-sharing](https://docs.docker.com/docker-for-mac/#file-sharing)
+        4. Share the LEAN root directory with docker
```
Comment (Martin-Molinero) [inline .idea/readme.md:46]: I think 1 and 2 wanted to be a hyperlink? not sure if it's working for some reason
image

edit: confirmed, it's a hyperlink for vscode

File: .idea/readme.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+<h4>Note About Python Algorithm Location</h4>
+
+
+Our specific configuration binds the Algorithm.Python directory to the container so any algorithm you would like to run must be in that directory. Please ensure your algorithm location looks just the same as the example above.
+
+<h3>Running Lean in the Container</h3>
+
+
+Now that your configuration is all set and your algorithm in the correct location use the run_docker.bat/.sh script located in Lean’s root directory. The script will kick off the process of starting the container.
```
Comment (Martin-Molinero) [inline .idea/readme.md:73]: nit - maybe can suggest using the PyCharm terminal

File: .idea/readme.md (hunk 4 above)
Comment (C-SELLERS) [reply .idea/readme.md]: Fixed!

File: .idea/readme.md (hunk 5 above)
Comment (C-SELLERS) [reply .idea/readme.md]: Done!

File: .dockerignore (hunk 3 above)
Comment (C-SELLERS) [reply .dockerignore:4]: Whoops was an early change but is no longer required. I will undo them.

Comment (C-SELLERS) [reply .dockerignore]: Done!

**Pull request merged**